File: Golang/Medium/1039_Minimum_Score_Triangulation_of_Polygon/solution.go
package minimumscoretriangulationofpolygon

import "math"

type state struct {
	i int
	j int
}

func minScoreTriangulation(values []int) int {
	return helper(values, 0, len(values)-1, make(map[state]int))
}

func helper(v []int, i int, j int, memo map[state]int) int {
	if j-i < 2 {
		return 0
	}

	key := state{i, j}
	if v, exist := memo[key]; exist {
		return v
	}

	minScore := math.MaxInt64
	for k := i + 1; k < j; k++ {
		score := v[i]*v[j]*v[k] + helper(v, i, k, memo) + helper(v, k, j, memo)
		minScore = min(minScore, score)
	}

	memo[key] = minScore
	return minScore
}

File: Python/Medium/1039_Minimum_Score_Triangulation_of_Polygon/solution.py
class Solution:
    def minScoreTriangulation(self, values: list[int]) -> int:
        return Solution.helper(values, 0, len(values)-1, {})
    
    def helper(v, i, j, memo):
        if j - i < 2:
            return 0

        key = (i, j)
        if key in memo:
            return memo[key]

        minScore = float("inf")
        for k in range(i + 1, j):
            score = (v[i] * v[j] * v[k]) + Solution.helper(v, i, k, memo) + Solution.helper(v, k, j, memo)
            minScore = min(minScore, score)
        
        memo[key] = minScore
        return minScore